File: [138] Copy List with Random Pointer/138.copy-list-with-random-pointer.cpp
/*
 * @lc app=leetcode id=138 lang=cpp
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) return nullptr;
        // no hashmap solution
        Node* current = head;
        Node* next = nullptr;

        // First round: make copy of each node,
        // and link them together side-by-side in a single list.
        while (current) {
            next = current->next;
            Node* copy = new Node(current->val);
            current->next = copy;
            copy->next = next;
            current = next;
        }
        // Second round: assign random pointers for the copy nodes.
        current = head;
        while (current) {
            if (current->random) {
                current->next->random = current->random->next;
            }
            current = current->next->next;
        }
        // Third round: restore the original list, and extract the copy list.
        Node* copyhead = head->next;
        current = head;
        while (current) {
            next = current->next->next;
            Node* copy = current->next;
            current->next = next;
            if (next) copy->next = next->next;
            current = next;
        }

        return copyhead;
    }
};
// @lc code=end

File: [138] Copy List with Random Pointer/138.copy-list-with-random-pointer.java
import java.util.HashMap;
import java.util.Hashtable;

/*
 * @lc app=leetcode id=138 lang=java
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;
        Map<Node, Node> ht = new HashMap<>();
        Node current = head;
        while (current != null) {
            // Create nodes copy
            if (!ht.containsKey(current)) {
                ht.put(current, new Node(current.val));
            }
            // Create random and next relation
            if (current.random != null && !ht.containsKey(current.random)) {
                ht.put(current.random, new Node(current.random.val));
            }
            if (current.next != null && !ht.containsKey(current.next)) {
                ht.put(current.next, new Node(current.next.val));
            }
            ht.get(current).random = ht.get(current.random);
            ht.get(current).next = ht.get(current.next);
            current = current.next;
        }
        return ht.get(head);
    }
}
// @lc code=end

